src/comp.c

```diff
@@ -0,0 +1,249 @@
+// [cheri-examples]: https://github.com/capablevms/cheri-examples/tree/d6ecb72aaa479f6126a69b9e70de1ec2fc49cdc0
+
+#include "comp.h"
+
+#include <elf.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <link.h>
+#include <dlfcn.h>
+#include <assert.h>
+
+#include <cheriintrin.h>
+
+#define __comp_mgr_data __attribute__((section(".data.cheri_comp_manager")))
+#define __comp_mgr_code __attribute__((section(".cheri_comp_manager")))
+#define MAX_SHNUM 256
+
+extern void __comp_mgr_code enter_compartment(void *sp, void *__capability comp_pcc,
+        void *__capability comp_ddc, void *__capability comp_entry);
+
+extern void __comp_mgr_code __comp_mgr_code_start(void);
+
+extern size_t __comp_mgr_code next_comp_id(void);
+extern void __comp_mgr_data_start(void);
+
+// A handle to the currently executing binary.
+static __comp_mgr_data FILE *ELF = NULL;
+// The ELF header of the binary.
+static __comp_mgr_data Elf64_Ehdr ELF_HDR = {0};
+// The cached seciton headers.
+static __comp_mgr_data Elf64_Shdr SHDRS[MAX_SHNUM] = {0};
+// The manager DDC and PCC.
+static __comp_mgr_data void *__capability DDC_PCC[2];
+// The sealed manager DDC and PCC.
+static __comp_mgr_data void *__capability SWITCHER_CAPS = NULL;
+// The manager PCC.
+static __comp_mgr_data void *__capability MGR_PCC = NULL;
+// The manager DDC.
+static __comp_mgr_data void *__capability MGR_DDC = NULL;
+
+// Compartment data area.
+// NOTE: This is currently only used for the compartment stack.
+static
+__attribute__((aligned(16))) char COMP_DATA[COMP_TOTAL_DATA_SIZE] = {0};
+
+// The address of `COMP_DATA`.
+//
+// This address must be stored within the boundaries of the compartment manager
+// data section during initialization, before restricting the DDC (this is
+// necessary because the address of `COMP_DATA` is loaded from a literal pool
+// located outside the managing compartment boundary).
+//
+// TODO: come up with a more readable workaround.
+static __comp_mgr_data void *__capability COMP_DATA_ADDR = NULL;
+
+static inline  __comp_mgr_code void
+seal_cap(void *__capability cap) {
+    __asm__("seal %w0, %w0, lpb" : "+r"(cap) :);
+}
+
+// Parse the section headers of the specified binary.
+static __comp_mgr_code int
+init_elf_info(char *bin_path) {
+    if ((ELF = fopen(bin_path, "r")) == NULL) {
+        perror("failed to open file");
+        return EXIT_FAILURE;
+    }
+
+    if (fread(&ELF_HDR, sizeof(ELF_HDR), 1, ELF) != 1) {
+        perror("failed to read ELF header");
+        return EXIT_FAILURE;
+    }
+
+    if (ELF_HDR.e_shnum > MAX_SHNUM) {
+        return EXIT_FAILURE;
+    }
+
+    for (size_t i = 0; i < ELF_HDR.e_shnum; ++i) {
+        if (fseek(ELF, ELF_HDR.e_shoff + i * sizeof(Elf64_Shdr), SEEK_SET)) {
+            return EXIT_FAILURE;
+        }
+
+        if (fread(&SHDRS[i], sizeof(Elf64_Shdr), 1, ELF) != 1) {
+            return EXIT_FAILURE;
+        }
+    }
```

Comment:
The `fseek` can be moved out of the loop

```suggestion
    if (fseek(ELF, ELF_HDR.e_shoff, SEEK_SET)) {
        return EXIT_FAILURE;
    }

    for (size_t i = 0; i < ELF_HDR.e_shnum; ++i) {

        if (fread(&SHDRS[i], sizeof(Elf64_Shdr), 1, ELF) != 1) {
            return EXIT_FAILURE;
        }
    }

```